lsp-mode.el

```diff
@@ -160,30 +196,50 @@ Optional arguments:
 
 `:initialize' is a function called when the client is initialized. It takes a
   single argument, the newly created client."
-  (let ((enable (intern (format "%s-enable" name))))
-    `(defun ,enable ()
-       ,docstring
-       (interactive)
-       (let ((client (make-lsp--client
-                       :language-id (if ,language-id-fn ,language-id-fn #'(lambda (_) ,language-id))
-                       :send-sync #'lsp--stdio-send-sync
-                       :send-async #'lsp--stdio-send-async
-                       :new-connection (lsp--make-tcp-connection
+  (let ((enable-interactive (intern (format "%s-enable" name))))
+    `(progn
+       (lsp-define-whitelist-enable ,name ,get-root)
+       (lsp-define-whitelist-disable ,name ,get-root)
+       (lsp-define-interactive-client ,name)
+
+       (defun ,enable-interactive (&optional force)
+         ,docstring
+         (interactive)
+         (let ((client (make-lsp--client
+                        :language-id (if ,language-id-fn ,language-id-fn #'(lambda (_) ,language-id))
+                        :send-sync #'lsp--stdio-send-sync
+                        :send-async #'lsp--stdio-send-async
+                        :new-connection (lsp--make-tcp-connection
                                          ,(symbol-name name)
                                          ,command
                                          ,command-fn
                                          ,host ,port)
-                       :get-root ,get-root
-                       :ignore-regexps ,ignore-regexps)))
-         (unless lsp-mode
-           ,(when initialize
-              `(funcall ,initialize client))
-           (let ((root (funcall (lsp--client-get-root client))))
-             (if (lsp--should-start-p root)
-               (progn
-                 (lsp-mode 1)
-                 (lsp--start client ,extra-init-params))
-               (message "Not initializing project %s" root))))))))
+                        :get-root ,get-root
+                        :ignore-regexps ,ignore-regexps)))
+           (lsp--start-client-maybe client force
+                                    :extra-init-params ,extra-init-params
+                                    :initialize ,initialize))))))
+
+(cl-defun lsp--start-client-maybe (client force
+                                 &key extra-init-params initialize)
+ "Start a CLIENT if lsp-mode is currently inactive, and the whitelist/blacklist allows it.
+If FORCE is `t', start regardless off whitelist/blacklist setting."
+            (unless lsp-mode
+              (when initialize
+                 (funcall initialize client))
+              (let ((root (funcall (lsp--client-get-root client))))
+                (if (lsp--should-start-p root)
```

Comment:
Nitpick: could be `(or force (lsp--should-start-p root)` and 
```
(lsp-mode 1)
(lsp--start client extra-init-params)
```
could be removed from the lower if

Reply:
Thanks, nothing like getting lessons in idiomatic elisp

Reply:
Can't say if that is idiomatic elisp, just my general programming skills that want to remove duplicate code :D